### PR TITLE
IMP: Adds method to raise an error when filtering results in an empty table

### PR DESCRIPTION
**Pull request by hagenjp**

### CODE REVIEW

File: q2_feature_table/plugin_setup.py [hunk 1]
```diff
@@ -379,7 +383,10 @@
                       'retained.'),
         'prevalence': ('The minimum portion of samples that a feature '
                        'must have a relative abundance of at least '
-                       '`abundance` to be retained.')
+                       '`abundance` to be retained.'),
+        'allow_empty_table': 'If true, the filtered table may be empty. '
+                             'Otherwise, an error will be raised if the '
```
Comment (lizgehret) [inline q2_feature_table/plugin_setup.py]:
```suggestion
                             'Default behavior is to raise an error if the '
```
Same suggestion for all three descriptions.

File: q2_feature_table/plugin_setup.py [hunk 2]
```diff
@@ -438,6 +446,9 @@
                        'table instead of being retained.',
         'filter_empty_samples': 'If true, drop any samples where none of the '
                                 'retained features are present.',
+        'allow_empty_table': 'If true, the filtered table may be empty. '
+                                'Otherwise, an error will be raised if the '
```
Comment (lizgehret) [inline q2_feature_table/plugin_setup.py]:
```suggestion
                                'Default behavior is to raise an error if the '
```

File: q2_feature_table/plugin_setup.py [hunk 3]
```diff
@@ -343,6 +344,9 @@
                        'table instead of being retained.',
         'filter_empty_features': 'If true, features which are not present in '
                                  'any retained samples are dropped.',
+        'allow_empty_table': 'If true, the filtered table may be empty. '
+                             'Otherwise, an error will be raised if the '
```
Comment (lizgehret) [inline q2_feature_table/plugin_setup.py]:
```suggestion
                             'Default behavior is to raise an error if the '
```
Small adjustment just to clarify what the default behavior is!

File: q2_feature_table/tests/filter/test_conditional_filter.py [hunk 4]
```diff
@@ -40,6 +40,48 @@ def test_conditional(self):
         npt.assert_array_equal(known.ids(axis='observation'),
                                test_.ids(axis='observation'))
 
+    def test_allow_empty_table_true(self):
+        # test True
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_conditional_filter.py]: I'm not seeing anywhere in this test where we confirm that this table is in fact empty - you have the `allow_empty_table` set to True, but it would be good to add an assertion with an 'expected' table that's empty that you're asserting is equal to the actual table that results from this method.

File: q2_feature_table/tests/filter/test_conditional_filter.py [hunk 5]
```diff
@@ -40,6 +40,48 @@ def test_conditional(self):
...
+        # test False
+        table = biom.Table(
+            data=np.array([[0,   0,  10,   0,   0],
+                           [250, 250, 140,  90, 150],
+                           [250,  25, 100, 200, 100],
+                           [0, 225, 250, 210, 250]]),
+            sample_ids=['A', 'B', 'C', 'D', 'E'],
+            observation_ids=['bat', 'cat', 'rat', 'a-tat-tat']
+            )
+        with self.assertRaises(ValueError):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_conditional_filter.py]: When adding these kinds of assertions, it's best to be as specific as possible - what I'd recommend doing instead is using an `assertRaisesRegex` that contains a segment of that specific ValueError's text. That way we know we're hitting the expected error branch!

File: q2_feature_table/tests/filter/test_filter_features.py [hunk 6]
```diff
@@ -45,12 +45,12 @@ def test_min_frequency(self):
...
         self.assertEqual(actual, expected)
 
-        # filter all
+        # filter all raising ValueError
         table = Table(np.array([[0, 1, 1], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_features(table, min_frequency=5)
-        self.assertTrue(actual.is_empty())
+        with self.assertRaisesRegex(ValueError, 'empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_features.py]: This is good! I would include a bit more text for the regex you're expecting to see within this value error though, just to make sure there's no chance of overlap with another existing value error. Same suggestion for all of your test cases that include this assertion!

File: q2_feature_table/tests/filter/test_conditional_filter.py (hunk 5 above)
Comment (hagenjp) [reply q2_feature_table/tests/filter/test_conditional_filter.py]: That makes sense! I will do that. Thank you!

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 7]
```diff
@@ -799,14 +799,31 @@ def test_combine_exclude_ids_and_features_filters(self):
...
-                                max_features=1)
+        with self.assertRaisesRegex(ValueError, 'empty'):
+            filter_samples(table,
+                           metadata=metadata,
+                           where=where,
+                           exclude_ids=True,
+                           min_features=1,
+                           max_features=1)
+
+    def test_allow_empty_table_true(self):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py]: This is great! For the rest of your tests, I'd make sure and add a test case like this one that asserts an empty table is allowed (for each method where this is used).

**Review by lizgehret — CHANGES_REQUESTED** [review]
Great job @hagenjp! Some comments inline, mainly on unit tests. Let me know if you have any questions or need clarification on anything!

File: q2_feature_table/tests/filter/test_filter_samples.py (hunk 7 above)
Comment (hagenjp) [reply q2_feature_table/tests/filter/test_filter_samples.py]: Just to clarify here: Each test where I filter all the samples and test raising the 'Empty table' error should also have a corresponding test where the empty table is accepted with the allow_empty_table=True. Is that right?

Comment (lizgehret) [reply q2_feature_table/tests/filter/test_filter_samples.py]: Yes, that's correct! You should be testing that both the true and false cases work as expected for each method where this new parameter is used.

File: q2_feature_table/tests/filter/test_conditional_filter.py [hunk 8]
```diff
@@ -40,6 +40,38 @@ def test_conditional(self):
         npt.assert_array_equal(known.ids(axis='observation'),
                                test_.ids(axis='observation'))
 
+    def test_allow_empty_table_true(self):
+        # test True
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_conditional_filter.py]:
```suggestion
```
Since this test method has a clearly defined test case that is reflected in the name, the comment isn't necessary here.

File: q2_feature_table/tests/filter/test_conditional_filter.py [hunk 9]
```diff
@@ -40,6 +40,38 @@ def test_conditional(self):
...
+            observation_ids=['bat', 'cat', 'rat', 'a-tat-tat']
+            )
+        test_ = filter_features_conditionally(table,
+                                              prevalence=0.9,
+                                              abundance=0.9,
+                                              allow_empty_table=True)
+        self.assertTrue(test_.is_empty())
+
+    def test_allow_empty_table_false(self):
+        # test False
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_conditional_filter.py]:
```suggestion
```
Same comment as the above True case.

File: q2_feature_table/tests/filter/test_filter_features.py [hunk 10]
```diff
@@ -190,6 +217,16 @@ def test_where(self):
                          ['S1', 'S2', 'S3'])
         self.assertEqual(actual, expected)
 
+    def test_allow_empty_table(self):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_features.py]: You can remove this test since we're testing the True case for this param with each of the above test cases!

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 11]
```diff
@@ -68,7 +68,14 @@ def test_min_frequency(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table, min_frequency=42)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:71]: I'd update the comment on line 67 to include 'raising ValueError' for consistency with the rest of your comments

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 12]
```diff
@@ -106,7 +113,13 @@ def test_max_frequency(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table, max_frequency=0)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:116]: Same comment as above re: the existing comment on line 112 ('raising ValueError')

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 13]
```diff
@@ -147,8 +160,16 @@ def test_filter_empty_features(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:163]: Same as above comments for comment on line 159

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 14]
```diff
@@ -176,7 +197,13 @@ def test_min_features(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table, min_features=3)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:200]: Same as above comments for comment on line 196

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 15]
```diff
@@ -204,7 +231,14 @@ def test_max_features(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table, max_features=0)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:234]: Same as above comments for comment on line 230

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 16]
```diff
@@ -242,7 +276,14 @@ def test_sample_metadata(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table, metadata=metadata)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:279]: Same as above comments for comment on line 273

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 17]
```diff
@@ -289,8 +330,18 @@ def test_sample_metadata(self):
                            'SampleType': ['gut', 'tongue', 'gut']},
                           index=pd.Index(['S1', 'S2', 'S3'], name='id'))
         metadata = qiime2.Metadata(df)
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:333]: Same as above comments for comment on line 328

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 18]
```diff
@@ -363,8 +414,8 @@ def test_where(self):
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
         where = "Subject='subject-1' AND Subject='subject-2'"
-        actual = filter_samples(table, metadata=metadata, where=where)
-        self.assertTrue(actual.is_empty())
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:417]: Same as above comments for comment on line 408 - additionally, you'll want to include the True case for this test as well.

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 19]
```diff
@@ -426,10 +477,26 @@ def test_where(self):
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
         where = "Subject='subject-1' OR Subject='subject-2'"
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:485]: Same as above comments for comment on line 471

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 20]
```diff
@@ -626,12 +693,12 @@ def test_combine_exclude_ids_and_frequency_filters(self):
         table = Table(np.array([[0, 1, 3], [1, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table,
-                                metadata=metadata,
-                                exclude_ids=True,
-                                max_frequency=4,
-                                min_frequency=3)
-        self.assertTrue(actual.is_empty())
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:701]: Same as above comments for comment on line 687/688 and you'll also want to include the True case for this test as well.

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 21]
```diff
@@ -644,12 +711,22 @@ def test_combine_exclude_ids_and_frequency_filters(self):
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
         where = "Subject='subject-1'"
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:728]: Same as above comments for comment on line 703-705

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 22]
```diff
@@ -781,12 +858,12 @@ def test_combine_exclude_ids_and_features_filters(self):
         table = Table(np.array([[0, 1, 3], [0, 1, 2]]),
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
-        actual = filter_samples(table,
-                                metadata=metadata,
-                                exclude_ids=True,
-                                min_features=1,
-                                max_features=1)
-        self.assertTrue(actual.is_empty())
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:875]: Same as above comments for comment on line 852/853, and also adding back in the True case for this test.

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 23]
```diff
@@ -799,14 +876,41 @@ def test_combine_exclude_ids_and_features_filters(self):
                       ['O1', 'O2'],
                       ['S1', 'S2', 'S3'])
         where = "SampleType='tongue'"
+        with self.assertRaisesRegex(ValueError, 'table is empty'):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py:900]: Same as above comments for comment on line 868-870

File: q2_feature_table/tests/filter/test_filter_samples.py [hunk 24]
```diff
@@ -799,14 +876,41 @@ def test_combine_exclude_ids_and_features_filters(self):
...
                                 metadata=metadata,
                                 where=where,
                                 exclude_ids=True,
                                 min_features=1,
-                                max_features=1)
+                                max_features=1,
+                                allow_empty_table=True)
         self.assertTrue(actual.is_empty())
 
+    def test_allow_empty_table_true(self):
```
Comment (lizgehret) [inline q2_feature_table/tests/filter/test_filter_samples.py]: This test (and the test below) can be removed since you're testing the True/False cases within all of the tests above.

**Review by lizgehret — CHANGES_REQUESTED** [review]
Okay I think this should be the final round of changes on this one - let me know if you have any questions on any of the requests. Great work!

**Review by lizgehret — CHANGES_REQUESTED** [review]
These changes look great! I think the _actual_ final thing here is just to remove the stray artifact directory that seems to have snuck into your latest commit @hagenjp 😉

**Comment by lizgehret** [comment]
This ci failure isn't on you @hagenjp - I'm updating the epoch for our next release (from the discussion about the release date in yesterday's eng meeting), and once that's done these actions should pass. No action needed on your end!

**Comment by hagenjp** [comment]
@lizgehret Oh I see. Thank you for telling me!

**Comment by lizgehret** [comment]
Great work! This is g2g 🙂

**Review by lizgehret — APPROVED** [review]

**Comment by lizgehret** [comment]
Hey @hagenjp final thing for you to do is update the description to include what this PR does. This makes our lives easier when adding things to the release changelog. 🙂
